content/blog/2025-05-13-brilooped/index.md

```diff
@@ -0,0 +1,443 @@
+        } else {
+            return NEXT;
+        }
+    } else {
+        return result;
+    }
+}
+
+case "while": {
+    while true {
+        if (!cond) {
+            return NEXT:
+        }
+        const result = evalBlock(children[0], state);
+        if (result.action == "break") {
+            if (result.level > 0) {
+                return { action: "break", level: result.level - 1 };
+            } else {
+                return NEXT;
+            }
+        } else if result.action == "continue" {
+            if (result.level > 0) {
+                return { action: "continue", level: result.level - 1 };
+            } else {
+                continue;
+            }
+        } else if (
+            result.action === "end" ||
+            result.action === "speculate" ||
+            result.action === "commit" ||
+            result.action === "abort"
+        ) {
+            return result;
+        }
+    }
+}
+
+```
+
+This evalBlock function was an additional evaluation helper function that was used to evaluate a block of instructions and return the resulting action. It was used to handle blocks of instructions in the children arrays:
+
+```typescript
+function evalBlock(instrs: bril.Instruction[], state: State): Action {
+  for (let i = 0; i < instrs.length; ++i) {
+    const instr = instrs[i];
+    if ("op" in instr) {
+      const action = evalInstr(instr, state);
+
+      switch (action.action) {
+        case "next":
+          break;
+        case "break":
+        case "continue":
+        case "end":
+        case "speculate":
+        case "commit":
+        case "abort":
+          return action;
+        default:
+          unreachable(action);
+          throw error(`unhandled action ${action.action}`);
+      }
+    }
+  }
+  return NEXT;
+}
+```
+
+The multi-level breaks and continues proved particularly challenging to implement, especially ensuring the correct bubble-up mechanism through nested control structures. Our solution hinged on the `level` counter, which elegantly tracks how many loop boundaries a `break` or `continue` statement needs to cross. During development, we encountered several cases where improper handling of the recursive evaluation caused programs to become trapped in infinite loops. These issues were difficult to debug — we ultimately resorted to execution profiling to trace the interpreter's control flow and identify where it was getting stuck. The resulting implementation not only handles complex control flow patterns correctly but does so with minimal additional overhead on the interpreter.
+
+## Beyond Relooper
+
+The Beyond Relooper algorithm was described by Ramsey as a way to translate Cmm sources into WebAssembly, for this project we adapted the algorith to translate Bril into Briloop. We'll start by providing a high level overview of the algorithm and some of the adaptations we did as the source and target languages are different from what are presented in Ramsey's paper.
```

Comment:
It might be nice to include a hyperlink or something to explain what Cmm is (I think it's an IR used in GHC?).